File: sequencer/software/sequencer.c
/**
 * SEQUENCER CONTROLLER
 * 
 * Authors:
 * Zach Baldwin
 * Andy Regli
 * Mark Vaughn
 *
 */

#include "hardware/gpio.h"

#include "serbus.h"
#include "sequencer.h"

// All of these can be modified by the ISRs
volatile unsigned int is_running;
volatile unsigned int is_mod_selected;
volatile unsigned int bpm;
volatile unsigned int time_sig;
volatile unsigned int current_beat;
repeating_timer_t timer;

int main ()
{
    // Hardware init
    stdio_init_all();

    // Initialize Sequencer states
    is_running = false;
    is_mod_selected = false;
    bpm = BPM_DEFAULT;
    time_sig = TS_DEFAULT;
    current_beat = 0;

    // Register ISRs to GPIO and Timer events
    // Tempo Encoder
    gpio_set_irq_enabled_with_callback(
        TEMPO_ENC0,
        GPIO_IRQ_EDGE_RISE | GPIO_IRQ_EDGE_FALL,
        true, // Enable interrupt
        &isr_tempo_encoder
        );
    gpio_set_irq_enabled_with_callback(
        TEMPO_ENC1,
        GPIO_IRQ_EDGE_RISE | GPIO_IRQ_EDGE_FALL,
        true, // Enable interrupt
        &isr_tempo_encoder
        );

    // Time signature encoder
    gpio_set_irq_enabled_with_callback(
        GPIO_ENC0,
        GPIO_IRQ_EDGE_RISE | GPIO_IRQ_EDGE_FALL,
        true, // Enable interrupt
        &isr_time_sig_encoder
        );
    gpio_set_irq_enabled_with_callback(
        TSIG_ENC1,
        GPIO_IRQ_EDGE_RISE | GPIO_IRQ_EDGE_FALL,
        true, // Enable interrupt
        &isr_time_sig_encoder
        );

    // Play / pause button
    gpio_set_irq_enabled_with_callback(
        PLYSTP_BTN,
        GPIO_IRQ_EDGE_FALL,
        true, // Enable interrupt
        &isr_play_pause
        );

    // Module selection status update
    gpio_set_irq_enabled_with_callback(
        MOD_STAT_IRQ,
        GPIO_IRQ_EDGE_RISE,
        true, // Enable interrupt
        &isr_module_status
        );

    // Set up the tick timer for micro beat generation    

    // Negative timeout means exact delay (rather than delay between callbacks)
    if (!add_repeating_timer_us((-1000000 * 60) / bpm, timer_callback, NULL, &isr_timer)) {
        // printf("Failed to add timer\n");
        // ERROR!
        return 1;
    }



    

    // Main event loop
    while (1) {
        update_screen();
        update_buttons();
    }
}


/**
 * Updates the screen. Displays the current tempo, time signature,
 * Play/Pause status, and relevant information during module programming
 */
void update_screen(void)
{
    // TODO
}


/**
 * Update the colors of the button matrix based on the current sequencer
 * state.
 */
void update_buttons(void)
{
    // TODO
}


/**
 * Handles generation of synchronization "tick" pulses with all modules
 */
bool isr_timer(repeating_timer_t *rt);
{
    // TODO!

    return true; // keep repeating    
}


/**
 * Handles detection of module status changes via the CBUS_EXTRA 
 * interrupt line
 * 
 * @param The GPIO pin number
 * @param The GPIO event type */
void isr_module_status(unsigned int gpio, uint32_t event)
{
    // TODO
}


/**
 * Handles updating of the global tempo value when the tempo encoder
 * is rotated
 * 
 * @param The GPIO pin number
 * @param The GPIO event type
 */
void isr_tempo_encoder(unsigned int gpio, uint32_t event)
{
    // TODO
}


/**
 * Handles the toggling of the is_running flag based on the play/pause
 * button input
 * 
 * @param The GPIO pin number
 * @param The GPIO event type
 */
void isr_play_pause(unsigned int gpio, uint32_t event)
{
    is_running = !is_running;
}


/**
 * Handles cycling of the time signature when the time signature 
 * encoder is rotated
 * 
 * @param The GPIO pin number
 * @param The GPIO event type
 */
void isr_time_sig_encoder(unsigned int gpio, uint32_t event)
{

}


// TODO: This might not be possible, depending on whether or not the pad can s
/**
 * Handle a button press on the squish pad
 * 
 * @param The GPIO pin number
 * @param The GPIO event type
 */
void isr_pad_event(void)
{
    // TODO
}

File: sequencer/software/sequencer.py
# from machine import Pin, Timer
# import busio
# from adafruit_neotrellis.neotrellis import NeoTrellis

########################################
# PINS
########################################

# PINS
GPIO_TP9          = 0
GPIO_ENC00        = 1
GPIO_ENC01        = 2
GPIO_TP10         = 3
GPIO_I2C0_SDA0    = 4
GPIO_I2C0_SCL0    = 5
GPIO_ENC10        = 6
GPIO_ENC11        = 7
# GPIO 8
# GPIO 9
GPIO_ENC20        = 10
GPIO_ENC21        = 11
GPIO_I2C0_SDA1    = 12
GPIO_I2C0_SCL1    = 13
GPIO_ENC30        = 14
GPIO_ENC31        = 15
GPIO_CBUS_EXTRA   = 16
GPIO_CBUS_DRDY    = 17
GPIO_CBUS_SCK     = 18
GPIO_CBUS_SDIN    = 19
GPIO_CBUS_SDOUT   = 20
# GPIO 21
GPIO_CBUS_NXTPR = 22
GPIO_BTN0       = 23
GPIO_BTN1       = 24
# GPIO 25
GPIO_ADC0       = 26
GPIO_ADC1       = 27
GPIO_ADC2       = 28
GPIO_ADC3       = 29


# Named pin mapping
TEMPO_ENC0      = GPIO_ENC00
TEMPO_ENC1      = GPIO_ENC01
TSIG_ENC0       = GPIO_ENC10
TSIG_ENC1       = GPIO_ENC11
PLYSTP_BTN      = GPIO_BTN0
MOD_STAT_IRQ    = GPIO_CBUS_EXTRA

# Tempo settings
BPM_MIN =     20
BPM_DEFAULT = 120
BPM_MAX =     600

# Time signature settings
TS_DEFAULT = 1

TIME_SIGNATURES = [
    [3, 4],
    [4, 4],
    [7, 8]
]


# Defaults
bpm = 120
is_running = False
is_mod_selected = False
time_sig = TS_DEFAULT
current_beat = 0
P_DRDY = Pin(GPIO_CBUS_DRDY, Pin.OUT)

# Callback structs
tick_timer = Timer()


def update_screen():
    pass

def update_buttons():
    pass

def isr_timer(timer):
    global P_DRDY
    P_DRDY.toggle()

def isr_module_status():
    pass

def isr_tempo_encoder():
    pass

def isr_play_pause():
    pass

def isr_time_sig_encoder():
    pass

def isr_pad_event():
    pass

def oled_init():
    pass

def pad_init():
    pass

# Init
oled_init();
pad_init();


# Set up the timer for micro beat ticks
tick_timer.init(freq=bpm/60, mode=Timer.PERIODIC, callback=isr_timer)


